### PR TITLE
e2e binds tests and nested bind fix

### LINKED ISSUE #4435: Add e2e tests for bind mount
Add e2e tests for bind issues (related to #4411)

### PR DESCRIPTION
**Pull request by cclerget**
## Description of the Pull Request (PR):

Add e2e bind tests and fix nested binds 

### This fixes or addresses the following GitHub issues:

 - Fixes #4435 

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: e2e/actions/actions.go [hunk 1]
```diff
@@ -1017,6 +1017,358 @@ func (c actionTests) actionNetwork(t *testing.T) {
...
+			args: []string{
+				"--writable",
+				"--bind", hostCanaryFile,
+				sandbox,
+				"test", "-f", hostCanaryFile,
+			},
+			exit: 0,
+		},
+		{
+			name: "SimpleFileWritableKO",
```
Comment (WestleyK) [inline e2e/actions/actions.go:1171]: Maybe `OK`?

```suggestion
			name: "SimpleFileWritableOK",
```

File: e2e/actions/actions.go [hunk 2]
```diff
@@ -1017,6 +1017,358 @@ func (c actionTests) actionNetwork(t *testing.T) {
...
+			args: []string{
+				"--writable",
+				"--bind", hostCanaryDir,
+				sandbox,
+				"test", "-f", hostCanaryFile,
+			},
+			exit: 0,
+		},
+		{
+			name: "SimpleDirWritableKO",
```
Comment (WestleyK) [inline e2e/actions/actions.go:1191]: Here too

```suggestion
			name: "SimpleDirWritableOK",
```

File: e2e/actions/actions.go [hunk 3]
```diff
@@ -1017,6 +1017,358 @@ func (c actionTests) actionNetwork(t *testing.T) {
...
+				"--bind", hostCanaryDir + ":/home/e2e/canaryRW",
+				"--writable",
+				sandbox,
+				"test", "-f", "/home/e2e/canaryRW/file",
+			},
+			postRun: checkHostDir(filepath.Join(hostHomeDir, "canaryRW")),
+			exit:    0,
+		},
+		{
+			name: "CustomHomeBindWritableKO",
```
Comment (WestleyK) [inline e2e/actions/actions.go:1269]:
```suggestion
			name: "CustomHomeBindWritableOK",
```

File: e2e/actions/actions.go [hunk 4]
```diff
@@ -1017,6 +1017,358 @@ func (c actionTests) actionNetwork(t *testing.T) {
...
+				sandbox,
+				"test", "-f", "/scratch/dir/file",
+			},
+			postRun: checkHostDir(filepath.Join(hostWorkDir, "scratch/scratch", "dir")),
+			exit:    0,
+		},
+	}
+
+	for _, profile := range e2e.Profiles {
+		profile := profile
```
Comment (WestleyK) [inline e2e/actions/actions.go:1353]: Is this profile variable assignment necessary here? Would the `profile` work without re-assigning it?

File: e2e/actions/actions.go [hunk 5]
```diff
@@ -1017,6 +1017,358 @@ func (c actionTests) actionNetwork(t *testing.T) {
...
+		createWorkspaceDirs(t)
+
+		t.Run(profile.String(), func(t *testing.T) {
```
Comment (WestleyK) [inline e2e/actions/actions.go:1356]: Is the `t.Run` necessary here? I thought `c.env.RunSingularity()` will take care of the `t.Run` stuff?

**Review by WestleyK — CHANGES_REQUESTED** [review]
Looks fine to me, just a couple comments/suggestions.

File: e2e/actions/actions.go (hunk 1 above)
Comment (cclerget) [reply e2e/actions/actions.go:1171]: As the test is expected to fail I put `KO` suffix

File: e2e/actions/actions.go (hunk 2 above)
Comment (cclerget) [reply e2e/actions/actions.go:1191]: same

File: e2e/actions/actions.go (hunk 4 above)
Comment (cclerget) [reply e2e/actions/actions.go:1353]: Yes it is, for details https://github.com/golang/go/wiki/CommonMistakes

File: e2e/actions/actions.go (hunk 5 above)
Comment (cclerget) [reply e2e/actions/actions.go:1356]: Yes `RunSingularity` could do that too, here this is just a convenient way to execute all tests for a profile within a parent test with profile name.

File: e2e/actions/actions.go (hunk 1 above)
Comment (WestleyK) [reply e2e/actions/actions.go:1171]: Okay, that makes sense.

**Review by WestleyK — APPROVED** [review]
Looks good to me :+1:

**Pull request merged**